### ISSUE TITLE
[Bug]: SFT samples longer than max_len train on no assistant tokens

### ISSUE THREAD

**Issue by mantreshkhurana** [Spyxpo/framerai#234]
### Preflight checklist

- [x] I searched existing issues and did not find a duplicate.
- [x] I am running the latest version of the affected branch.

### Affected component

Training / data pipeline

### Description

`ChatTemplate.encode_conversation` truncates from the front. At
`model/tokenizer/chat_template.py:158-160` an over-long sample is cut with `input_ids[:max_len]`
and `labels[:max_len]`, keeping the head of the conversation and discarding the tail.

The tail is where the assistant turn is. Labels are masked to the assistant spans at
`model/tokenizer/chat_template.py:156`, so a conversation whose final assistant turn falls past
`max_len` produces a label row that is entirely `-100`.

`model/framer.py:446` computes the text loss with `F.cross_entropy(..., ignore_index=-100)` at the
default `reduction="mean"`. With every target ignored the mean is taken over an empty set and the
loss is `nan`. The backward pass propagates it into every parameter the step touches and the
optimizer writes it into the weights, so a single over-long sample destroys the run. Nothing in
sample construction or in the training loop rejects or reports the condition.

`SFTDataset` and `DPODataset` both default `max_len` to 512 (`model/data.py:155`,
`model/data.py:216`), which an ordinary multi-turn record exceeds, so this is reached by normal
data rather than by an edge case.

The same function pads without producing an attention mask (`model/tokenizer/chat_template.py:161-164`
returns only `input_ids` and `labels`), so pad positions are attended as if they were content.

### Steps to reproduce

1. Build a JSONL SFT record whose user turn alone tokenises past `max_len`, followed by a short
   assistant turn.
2. Load it with `SFTDataset(data_dir, tokenizer, max_len=512)`.
3. Inspect the sample: `labels` is `-100` at every position.
4. Run one training step over a batch containing that sample.
5. `results["text_loss"]` is `nan`, and after `optimizer.step()` the model parameters are `nan`.

### Expected behavior

An over-long conversation should be truncated so the assistant turn survives, dropping the oldest
turns rather than the newest, matching the generation path, which trims a prompt keeping its end.

A sample with no unmasked target should be rejected at construction with a message naming the
record, rather than emitted and trained on, and a batch that somehow contains no targets should
be reported rather than written into the weights as `nan`.

`encode_conversation` should return an `attention_mask` alongside `input_ids` and `labels`, since
it pads.

### Logs and error output

```shell
>>> sample = dataset[0]
>>> (sample["labels"] != -100).sum()
tensor(0)

>>> out = model(input_ids=sample["input_ids"][None], labels=sample["labels"][None])
>>> out["text_loss"]
tensor(nan, grad_fn=<NllLossBackward0>)

Step 1/1000 | loss nan | lr 1.00e-06 | 4.1 it/s
```

### Environment

OS: macOS 15.6
Python: 3.12.4
PyTorch: 2.8.0 (CPU)
Node: 22.11.0
GPU/VRAM: not applicable
Affected branch: stable

**Comment by kunaldeo001** [comment]
I’d like to take this one up.

**Comment by mantreshkhurana** [comment]
Fixed by #246, now on stable.
